### PR TITLE
feat: support oneOf in messages

### LINKED ISSUE #62: Using oneOf, generator failing looking for ((channel["publish"])["message"])["name"] which is undefined or falsey
#### Describe the bug
When a channel uses `oneOf` instead of a single message, the generator fails
with Error: Unable to call `the return value of (the return value of (channel["publish"])["message"])["name"]`, which is undefined or falsey

#### How to Reproduce

Copy the sample from https://www.asyncapi.com/docs/tutorials/streetlights but use oneOf
(as a precursor to publishing multiple events to that channel).

Source:

```yaml
asyncapi: '2.0.0'
info:
  title: Streetlights API
  version: '1.0.0'
  description: |
    The Smartylighting Streetlights API allows you
    to remotely manage the city lights.
  license:
    name: Apache 2.0
    url: 'https://www.apache.org/licenses/LICENSE-2.0'
servers:
  mosquitto:
    url: mqtt://test.mosquitto.org
    protocol: mqtt
channels:
  light/measured:
    publish:
      summary: Inform about environmental lighting conditions for a particular streetlight.
      operationId: onLightMeasured
      message:
        oneOf:
          - name: LightMeasured
            payload:
              type: object
              properties:
                id:
                  type: integer
                  minimum: 0
                  description: Id of the streetlight.
                lumens:
                  type: integer
                  minimum: 0
                  description: Light intensity measured in lumens.
                sentAt:
                  type: string
                  format: date-time
                  description: Date and time when the message was sent.

```

run:

```bash
ag --output __sdk asyncapi.yaml @asyncapi/nodejs-template -p server=mosquitto

Something went wrong:
Template render error: (~/.nvm/versions/node/v12.21.0/lib/node_modules/@asyncapi/nodejs-template/template/src/api/routes/$$channel$$.js) [Line 13, Column 81]
  Error: Unable to call `the return value of (the return value of (channel["publish"])["message"])["name"]`, which is undefined or falsey
    at Object._prettifyError (~/dev/asyncapi/async-events-api-def/node_modules/nunjucks/src/lib.js:36:11)
    at ~/dev/asyncapi/async-events-api-def/node_modules/nunjucks/src/environment.js:563:19
    at Template.root [as rootRenderFunc] (eval at _compile (~/dev/asyncapi/async-events-api-def/node_modules/nunjucks/src/environment.js:633:18), <anonymous>:60:3)
    at Template.render (~/dev/asyncapi/async-events-api-def/node_modules/nunjucks/src/environment.js:552:10)
    at Environment.renderString (~/dev/asyncapi/async-events-api-def/node_modules/nunjucks/src/environment.js:380:17)
    at ~/dev/asyncapi/async-events-api-def/node_modules/@asyncapi/generator/lib/renderer/nunjucks.js:32:14
    at new Promise (<anonymous>)
    at nunjucksExport.renderNunjucks (~/dev/asyncapi/async-events-api-def/node_modules/@asyncapi/generator/lib/renderer/nunjucks.js:31:10)
    at Generator.renderFile (~/dev/asyncapi/async-events-api-def/node_modules/@asyncapi/generator/lib/generator.js:688:12)
    at Generator.renderAndWriteToFile (~/dev/asyncapi/async-events-api-def/node_modules/@asyncapi/generator/lib/generator.js:602:27)
```

#### Expected behavior

iterator over all the messages in the `oneOf`

### PR DESCRIPTION
**Pull request by DavidBiesack**
updated `$$channel$$.js` (and `lib/message-validator.js`) to support `oneOf:` messages

Modified the `validateMessage` function to validate the payload and if
valid, increment and return a counter of validated messages. Ignore
validation exceptions. Throw an exception only if `nValidated > 1` (no
need to try validating more - fail fast). `router` functions (publish
or subscribe) try to validate the payload against all of the messages.
If exactly one matches, then we proceed, else we stop.

I used this sample input to test code gen with both subscribe and publish

```yaml
asyncapi: '2.0.0'
info:
  title: Streetlights API
  version: '1.0.0'
  description: |
    The Smartylighting Streetlights API allows you
    to remotely manage the city lights.
  license:
    name: Apache 2.0
    url: 'https://www.apache.org/licenses/LICENSE-2.0'
servers:
  mosquitto:
    url: mqtt://test.mosquitto.org
    protocol: mqtt
channels:
  light/measured:
    publish:
      summary: Inform about environmental lighting conditions for a particular streetlight.
      operationId: pubLightMeasured
      message:
        oneOf:
          - name: LightMeasured
            payload:
              $ref: '#/components/schemas/LightMeasured'
          - name: LightFailure
            payload:
              $ref: '#/components/schemas/LightFailure'
    subscribe:
      summary: Inform about environmental lighting conditions for a particular streetlight.
      operationId: subLightMeasured
      message:
        oneOf:
          - name: LightMeasured
            payload:
              $ref: '#/components/schemas/LightMeasured'
          - name: LightFailure
            payload:
              $ref: '#/components/schemas/LightFailure'
components:
  schemas:
    LightMeasured:
      type: object
      required:
        - id
        - lumens
        - sentAt
      additionalProperties: false
      properties:
        id:
          type: integer
          minimum: 0
          description: Id of the streetlight.
        lumens:
          type: integer
          minimum: 0
          description: Light intensity measured in lumens.
        sentAt:
          type: string
          format: date-time
          description: Date and time when the message was sent.
    LightFailure:
      type: object
      required:
        - id
        - failedAt
        - sentAt
      additionalProperties: false
      properties:
        id:
          type: integer
          minimum: 0
          description: Id of the streetlight.
        failedAt:
          type: string
          format: date-time
          description: Date and time when the failure was detected.
        sentAt:
          type: string
          format: date-time
          description: Date and time when the message was sent.

```

This is the generated output from running
`ag --output sdk/nodejs asyncapi.yaml @asyncapi/nodejs-template -p server=mosquitto` :

```javascript
const Router = require('hermesjs/lib/router');
const {validateMessage} = require('../../lib/message-validator');
const router = new Router();
const lightMeasuredHandler = require('../handlers/light-measured');
module.exports = router;



/**
 * Inform about environmental lighting conditions for a particular streetlight.
 */
router.use('light/measured', async (message, next) => {
  try {
    
    let nValidated = 0;
    // For oneOf, only one message schema should match.
    // Validate payload against each message and count those which validate
    nValidated = await validateMessage(message.payload,'light/measured','LightMeasured','publish', nValidated);
    nValidated = await validateMessage(message.payload,'light/measured','LightFailure','publish', nValidated);
    if (nValidated === 1) {
      await lightMeasuredHandler.pubLightMeasured({message});
      next()
    } else {
      throw new Error(`${nValidated} of 2 message schemas matched when exactly 1 should match`);
    }
    
  } catch (e) {
    next(e);
  }
});
/**
 * Inform about environmental lighting conditions for a particular streetlight.
 */
router.useOutbound('light/measured', async (message, next) => {
  try {
    
    let nValidated = 0;
    // For oneOf, only one message schema should match.
    // Validate payload against each message and count those which validate
    nValidated = await validateMessage(message.payload,'light/measured','LightMeasured','subscribe', nValidated);
    nValidated = await validateMessage(message.payload,'light/measured','LightFailure','subscribe', nValidated);
    if (nValidated === 1) {
      await lightMeasuredHandler.subLightMeasured({message});
      next()
    } else {
      throw new Error(`${nValidated} of 2 message schemas matched when exactly 1 should match`);
    }
    
  } catch (e) {
    next(e);
  }
});
```
Possible enhancements: track which schemas validated to give better response to the caller via the exception.

**Related issue(s)**

Fixes #62

### CODE REVIEW

File: template/src/api/routes/$$channel$$.js [hunk 1]
```diff
@@ -11,9 +13,24 @@ module.exports = router;
   {%- endif %}
 router.use('{{ channelName | toHermesTopic }}', async (message, next) => {
   try {
+    {% if channel.publish().hasMultipleMessages() %}
```
Comment (derberg) [inline template/src/api/routes/$$channel$$.js:16]: the only problem I see is that the generated code is the same for subscribe and publish, except of the operation. You have 2 options:

Option 1. refactor this way that generated code is duplicated (in the end it is generated code so it doesn't matter) but the nunjucks code is also duplicated and should be extracted into a reusable `macro` in `partials` directory - example https://github.com/asyncapi/template-for-generator-templates/blob/nunjucks/partials/listChannels.html
Option 2. refactor this in a way that generated code is not duplicated but is actually using a helper function. I think Option 1 would be easier as writing such helper with nunjucks would add more complexity to template code

Your decision

Comment (DavidBiesack) [reply template/src/api/routes/$$channel$$.js:16]: I suggest deferring that refactoring to the future. Both are beyond my skillset. I would rather add some testing for this in some way; I don't actually know how we can validate this change (i.e. test the generated code) to ensure it is working.

Comment (DavidBiesack) [reply template/src/api/routes/$$channel$$.js:16]: Yes, that is exactly what I'm saying. The use of `oneOf` in AsyncAPI maps directly to a JSON schema that is defined with `oneOf`, so AsyncAPI should export that schema object for validation purposes.

Comment (derberg) [reply template/src/api/routes/$$channel$$.js:16]: oh, interesting, so `channel.subscribe().message()` would return a `oneOf` JSON Schema?

**Comment by DavidBiesack** [comment]
oh, interesting, so channel.subscribe().message() would return a oneOf JSON
Schema?


yes, exactly. And if AsyncAPI evolves to allow anyOf or oneOf here as well
as oneOf, all would work because this would just be a valid JSON schema.

**Comment by derberg** [comment]
@DavidBiesack would you mind creating an issue about it in the parser? and also add a TODO comment in the code that once it is implemented we need to refactor. We would proceed with this PR and merge and long term discuss your proposal. How does that sound?

**Comment by DavidBiesack** [comment]
@derberg yes, I'll try to do those; it may be a while before I can carve off some time.

**Comment by DavidBiesack** [comment]
I see a conflict but cannot resolve it in GitHub. [Resolve button](https://github.com/asyncapi/nodejs-template/pull/82/conflicts) is disabled.

**Review by derberg — APPROVED** [review]
@DavidBiesack hey, sorry David for such a huge delay. It got completely lost in my notifications. It looks great, and thanks a million for the issue in the Parser and detailed comment in code. 

I'm ok to merge. Since it was some time already, please confirm if all is good on your side

btw I changed to `feat` as this should go out as minor, as this is new feature from my point of view

**Comment by DavidBiesack** [comment]
@derberg thoughts on that failing test? that's deep in an area I'm not familiar with.

**Comment by derberg** [comment]
@DavidBiesack it is failing because new snapshot of generated JS files do not match what is there in the repo.

in your PR you modified `$$channel$$.js` so snapshot changed, as generated code looks different.

so for example in https://github.com/DavidBiesack/nodejs-template/blob/fix/issue-62/test/__snapshots__/integration.test.js.snap#L120-L130 rounter is
```js
router.useOutbound('smartylighting/streetlights/1/0/action/:streetlightId/dim', async (message, next) => {
  try {
    await validateMessage(message.payload,'smartylighting/streetlights/1/0/action/{streetlightId}/dim','dimLight','subscribe');
    await smartylightingStreetlights10ActionStreetlightIdDimHandler.dimLight({message});
    next();
  } catch (e) {
    next(e);
  }
});
"
`;
```

but generated content contains your comment added about issue in the parser

I think all will be fixed once you move this comment-> https://github.com/asyncapi/nodejs-template/pull/82/files#diff-c144fb671ad194876a8af5d422b103159158938d02ceedecb992ff78d363241fR16-R19 inside the `if`. So in generated code it will appear only if there are multiple messages in the asyncapi document

**Comment by DavidBiesack** [comment]
@derberg I changed the comment from `//` style to `/* ... */` so it is not written to the generated source.
I also added a `package.json` script to re-capture the snapshots:
```bash
npm run test.updateSnapshot
```

**Comment by derberg** [comment]
@DavidBiesack one fix more for the linter error needed

File: package.json [hunk 2]
```diff
@@ -20,6 +20,7 @@
   "homepage": "https://github.com/asyncapi/nodejs-template#readme",
   "scripts": {
     "test": "rimraf test/temp && jest --modulePathIgnorePatterns='./template'",
+    "test.updateSnapshot": "rimraf test/temp && jest --updateSnapshot --modulePathIgnorePatterns='./template'",
```
Comment (derberg) [inline package.json:23]:
```suggestion
    "test:updateSnapshot": "rimraf test/temp && jest --updateSnapshot --modulePathIgnorePatterns='./template'",
```

just to be consistent with other scripts

**Comment by DavidBiesack** [comment]
@derberg I merged the linter fix, and also your `test:updateSnapshot` proposal.

**Review by derberg — APPROVED** [review]
LGTM 🚀 

and sorry for keeping it so long on a PR 🙏🏼

**Pull request merged**